pkg/flow/internal/controller/node_config_import.go

```diff
@@ -0,0 +1,328 @@
+}
+
+// Evaluate implements BlockNode and evaluates the import source.
+func (cn *ImportConfigNode) Evaluate(scope *vm.Scope) error {
+	return cn.source.Evaluate(scope)
+}
+
+// onContentUpdate is triggered every time the managed import source has new content.
+func (cn *ImportConfigNode) onContentUpdate(importedContent string) {
+	cn.mut.Lock()
+	defer cn.mut.Unlock()
+
+	cn.inContentUpdate.Store(true)
+	defer cn.inContentUpdate.Store(false)
+
+	// If the source sent the same content, there is no need to reload.
+	if cn.importedContent == importedContent {
+		return
+	}
+
+	cn.importedContent = importedContent
+	cn.importedDeclares = make(map[string]ast.Body)
+	cn.importConfigNodesChildren = make(map[string]*ImportConfigNode)
+
+	parsedImportedContent, err := parser.ParseFile(cn.label, []byte(importedContent))
+	if err != nil {
+		level.Error(cn.logger).Log("msg", "failed to parse file on update", "err", err)
+		return
+	}
+
+	// populate importedDeclares and importConfigNodesChildren
+	cn.processImportedContent(parsedImportedContent)
+
+	// evaluate the importConfigNodesChildren that have been created
+	err = cn.evaluateChildren()
+	if err != nil {
+		level.Error(cn.logger).Log("msg", "failed to evaluate nested import", "err", err)
+		return
+	}
+
+	// trigger to stop previous children from running and to start running the new ones.
+	if cn.importChildrenRunning {
+		cn.importChildrenUpdateChan <- struct{}{}
+	}
+
+	cn.OnBlockNodeUpdate(cn)
+}
+
+// processImportedContent processes declare and import blocks of the provided ast content.
+func (cn *ImportConfigNode) processImportedContent(content *ast.File) {
+	for _, stmt := range content.Body {
+		switch stmt := stmt.(type) {
+		case *ast.BlockStmt:
+			fullName := strings.Join(stmt.Name, ".")
+			switch fullName {
+			case "declare":
+				cn.processDeclareBlock(stmt)
+			case importsource.BlockImportFile, importsource.BlockImportString: // TODO: add other import sources
+				cn.processImportBlock(stmt, fullName)
+			default:
+				level.Error(cn.logger).Log("msg", "only declare and import blocks are allowed in a module", "forbidden", fullName)
+			}
+		default:
+			level.Error(cn.logger).Log("msg", "only declare and import blocks are allowed in a module")
+		}
+	}
+}
+
+// processDeclareBlock stores the declare definition in the importedDeclares.
+func (cn *ImportConfigNode) processDeclareBlock(stmt *ast.BlockStmt) {
+	if _, ok := cn.importedDeclares[stmt.Label]; ok {
+		level.Error(cn.logger).Log("msg", "declare block redefined", "name", stmt.Label)
+		return
+	}
+	cn.importedDeclares[stmt.Label] = stmt.Body
+}
+
+// processDeclareBlock creates an ImportConfigNode child from the provided import block.
+func (cn *ImportConfigNode) processImportBlock(stmt *ast.BlockStmt, fullName string) {
+	sourceType := importsource.GetSourceType(fullName)
+	if _, ok := cn.importConfigNodesChildren[stmt.Label]; ok {
+		level.Error(cn.logger).Log("msg", "import block redefined", "name", stmt.Label)
+		return
```

Comment:
Should this error be raised up higher? Should this cause something else to fail rather than being constrained to logs?